Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (C) 2024 Igalia S.L
+ * Copyright (C) 2024 Metrological Group B.V.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * aint with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "GStreamerQuirkBroadcom.h"
+
+#include <string.h>
+
+#if USE(GSTREAMER)
+
+#include "GStreamerCommon.h"
+
+namespace WebCore {
+
+GST_DEBUG_CATEGORY_STATIC(webkit_broadcom_base_quirks_debug);
+#define GST_CAT_DEFAULT webkit_broadcom_base_quirks_debug
+
+GStreamerQuirkBroadcomBase::GStreamerQuirkBroadcomBase()
+{
+    GST_DEBUG_CATEGORY_INIT(webkit_broadcom_base_quirks_debug, "webkitquirksbroadcombase", 0, "WebKit Broadcom Base Quirks");
+}
+
+bool GStreamerQuirkBroadcomBase::queryBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, const char*& elementName, GRefPtr<GstQuery>& query) const
+{
+    if (!isEnsuredOwnedState(playerPrivate))
+        return false;
+    GStreamerQuirkBroadcomBaseState& state = static_cast<GStreamerQuirkBroadcomBaseState&>(playerPrivate->quirkState());
+
+    if (playerPrivate->shouldDownload() || !state.m_queue2)
+        return false;
+    if (gst_element_query(state.m_queue2.get(), query.get())) {
+        elementName = "queue2";
+        return true;
+    }
+    return false;
+}
+
+int GStreamerQuirkBroadcomBase::correctBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, int originalBufferingPercentage, GstBufferingMode mode) const
+{
+    if (!isEnsuredOwnedState(playerPrivate))
+        return originalBufferingPercentage;
+
+    GStreamerQuirkBroadcomBaseState& state = static_cast<GStreamerQuirkBroadcomBaseState&>(playerPrivate->quirkState());
+
+    // The Nexus playpump buffers a lot of data. Let's add it as if it had been buffered by the GstQueue2
+    // (only when using in-memory buffering), so we get more realistic percentages.
+    if (mode != GST_BUFFERING_STREAM || !state.m_vidfilter)
```

Comment:
I'm going to review myself here to avoid forgetting about [a problem detected downstream](https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1309#issuecomment-2223461375):

> The Broadcom buffer percentage correction isn't applied in the case of Deezer because that is an audio-only service, and the code looks for vidfilter (not present). I can work on trying to apply the same corrections when only audfilter is present.

So: Maintain references to m_audfilter, check if any of m_vidfilter or m_audfilter is present, and ask for buffered-bytes to the first of them. Also, the property is declared in the [original source code](https://github.com/Metrological/bcm-collaboration/blob/master/reference/vidfilter/src/gst_brcm_vid_filter.c#L452) as "buffered_bytes" (with underscore), instead of "buffered-bytes" (with minus sign). I would swear it worked anyway, but I have to double-check.